src/dogroups.c

```diff
@@ -435,6 +438,19 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
         if (copied) UNPROTECT(1);
       }
     }
+    // progress printing, #3060
+    // could potentially refactor to use fread's progress() function, however we would lose some information in favor of simplicity.
+    double now;
+    if (showProgress && (now=wallclock())>=nextTime) {
+      double avgTimePerGroup = (now-startTime)/(i+1);
+      int ETA = (int)(avgTimePerGroup*(ngrp-i-1));
+      if (hasPrinted || ETA >= 0) {
+        if (verbose && !hasPrinted) Rprintf(_("\n"));
+        Rprintf(_("\rProcessed %d groups out of %d. %.0f%% done. Time elapsed: %ds. ETA: %ds."), i+1, ngrp, 100.0*(i+1)/ngrp, (int)(now-startTime), ETA);
```

Comment:
why is there inconsistency on whether the message ends with `\n`?

Reply:
If I'm correct you are talking about lines 449 and 462? 449 doesn't include `\n` because it is updated many times and I would like for it to replace the previous line instead of printing to a new line. 462 includes `\n` because things printed afterwards need to be in a new line (later verbose messages, resulting data.table, etc). I believe `fread`'s `progress` does something similar where the the final call at the end of the operation prints a complete output with a new line

Reply:
If we're worried about this inconsistency I think we can get around it by deferring to a helper